Refactoring the `Qasm3Visitor`

TheGupta2012 opened the pull request:
Fixes #136 
<!--
Before submitting a pull request, please read:

- https://github.com/qBraid/qbraid-qir/blob/main/CONTRIBUTING.md#pull-requests

⚠️ Your pull request title should be short, detailed, and understandable for all.
⚠️ Please link any issues that this PR aims to close, if applicable.
⚠️ If you believe this PR should be highlighted in the qBraid-QIR CHANGELOG, please add a new entry to the `CHANGELOG.md` file, summarizing the change, and including a link back to the PR.
-->

## Summary of changes
- With the addition of new constructs in the visitor, a refactoring was needed for the module 
- Validators, helpers and some other methods were factored out of the `Qasm3Visitor`. These were independent of the `self` and moved out to a separate utility class. Moreover, it made sense to keep all the `_visit_*` methods in the visitor.
- Imports for `openqasm3` were also moved to a different file
- One possible separation could also be done based on the constructs i.e. `GateUtils`, `VariableUtils`, etc. but felt a bit of an overkill

@ryanhill1 would like to hear your thoughts on this too, thanks!

ryanhill1 commented on qbraid_qir/qasm3/utils/imports.py:
A dedicated file for imports isn't great practice. A better way to condense import statements in a given file is to use namespace imports instead of explicit imports. For example, instead of: 

```python
from openqasm3.ast import (
    AliasStatement,
   ...
)


if instance(stmnt, AliasStatement):
    ...

```

this would become

```python
import openqasm3.ast as ast

if instance(stmnt, ast.AliasStatement):
    ...
```

- **Explicit imports** are useful when you need only a few specific items from a module, and it's clear from the context what they are.
- **Namespace imports** are beneficial when working with large modules, and to avoid conflicts with local names or when you want clarity about the origin of each used component.

ryanhill1 commented on qbraid_qir/qasm3/visitor.py:
```diff
@@ -14,81 +14,31 @@
 from .exceptions import Qasm3ConversionError
-from .oq3_maps import (
+from .utils.expressions import Qasm3ExprEvaluator
+from .utils.imports import *
```
Instead of doing a wildcard import of a module filled with explicit imports, let's just use a namespace import. See comment above.

ryanhill1 reviewed (changes requested):
1. Let's remove the `qasm3.utils.imports` file and instead use namespace imports in `qasm3.visitor` and other files that have big import blocks.
2. Please refactor to eliminate any ascending relative imports. This can be done either by  moving all of files within `qasm3.utils` up to the `qasm3` module level, or potentially with some clever inversion of the module structure. In the former, there would still only be 8 python files within the top-level `qasm3` module, which isn't that bad.

See comments below.

TheGupta2012 commented on qbraid_qir/qasm3/utils/maps.py:
```diff
@@ -32,9 +32,9 @@
-from .linalg import kak_decomposition_angles
+from ..elements import InversionOp
+from ..exceptions import Qasm3ConversionError
+from ..linalg import kak_decomposition_angles
```
I think having the utilities at the base level is fine too, 8 files aren't much

ryanhill1 reviewed (changes requested):
Overall looking much better! Last few things to address which I mention in more detail below:

- Making sure all functions have appropriate type hints (this will also help make  https://github.com/qBraid/qbraid-qir/issues/145 easier)
- Combine in the `print_err_location` and exception raising by either modifying the existing `Qasm3ConversionError` exception class or creating a helper function that both raises the exception and logs/prints the element line/column traceback.
- Rename the `visitor_utils` file and `Qasm3VisitorUtils` class to something a bit more descriptive. Regroup validators / analyzers / transformers as necessary.

TheGupta2012 commented on qbraid_qir/qasm3/visitor_utils.py:
```diff
@@ -0,0 +1,566 @@
+from .maps import LIMITS_MAP, VARIABLE_TYPE_MAP, qasm_variable_type_cast
+
+
+class Qasm3VisitorUtils:
```
Okay got it

TheGupta2012 commented on qbraid_qir/qasm3/visitor_utils.py:
```diff
@@ -0,0 +1,566 @@
+
+    # ************* Generic utilities *************
+    @staticmethod
+    def print_err_location(element: Span) -> None:
```
I agree that the approach to create a helper method would reduce redundancy.

TheGupta2012 commented:
- Hi @ryanhill1 , I have incorporated the comments and segregated the validator, transformer and analyser into respective files. 
- I have also changed the structure of these classes to include an [attribute which refers to the `BasicQasmVisitor` object](https://github.com/qBraid/qbraid-qir/pull/142/commits/80662444a5c8ad8d61e4057c78f65c5ff60d3ccd#diff-de65fad9d5e9503ce107d80cf4435b642e6edc96da50ed8ebdeb560a2cfaddb0R37). This will remove the dependency of the from the function calls. Methods referring to the object are now `classmethod`s and reference this object internally.

ryanhill1 commented on qbraid_qir/qasm3/analyzer.py:
```diff
@@ -0,0 +1,184 @@
+                )
+            return condition.rhs.value != 0
+        if not isinstance(condition, IndexExpression):
+            raise_qasm3_error(
+                message=(
+                    f"Unsupported expression type '{type(condition)}' in if condition. "
+                    "Can only be a simple comparison"
+                ),
+                err_type=Qasm3ConversionError,
+                span=condition.span,
+            )
```
Based on the type hint `condition: Union[UnaryExpression, BinaryExpression, IndexExpression]`, this code is unreachable.

ryanhill1 commented on qbraid_qir/qasm3/transformer.py:
```diff
@@ -0,0 +1,212 @@
+            return condition.lhs.index[0].value, condition.lhs.collection.name
+        if isinstance(condition, IndexExpression):
+            return condition.index[0].value, condition.collection.name
+        return None, None
```
Based on the type hint `condition: Union[UnaryExpression, BinaryExpression, IndexExpression]`, this line is unreachable

ryanhill1 reviewed (changes requested):
These updates look great! If you could just address the two lines of unreachable code I noted below either by updating the typehint, or removing them. After that it should be good

*Edit*: Also, is the [supported conversions table](https://github.com/qBraid/qbraid-qir/tree/main/qbraid_qir/qasm3#supported-conversions-status-table) up to date? And if not, could you update it? Thanks!

TheGupta2012 commented:
Sure, will change. Also, some type hints are still left but I'll address them in the `mypy` issue
